### PR TITLE
Implement @file-services/cached

### PR DESCRIPTION
**Pull request by tomermes**

### CODE REVIEW

**Comment by tomermes** [comment]
@AviVahl Added ix tests

File: packages/cached/package.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  "scripts": {
+    "clean": "rimraf ./cjs ./esm",
+    "build": "ts-build ./src --cjs --esm",
+    "test": "run-p test:node test:browser",
+    "test:node": "mocha \"test/**/*.{spec,ix}.ts?(x)\"",
+    "test:browser": "mocha-pup \"test/**/*.spec.ts?(x)\"",
+    "prepack": "yarn build"
+  },
+  "dependencies": {
+    "@file-services/posix-path": "^1.0.3",
```
Comment (AviVahl) [inline packages/cached/package.json]: remove, not used.

File: packages/cached/test/cached-fs-node.ix.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
+import { expect } from 'chai';
+import { createNodeFs } from '@file-services/node';
+import { createCachedFs } from '../src';
+
+describe(`cachedFs with Node's fs`, () => {
+    it('caches statsSync calls', async () => {
+        const filePath = '../../package.json';
```
Comment (AviVahl) [inline packages/cached/test/cached-fs-node.ix.ts]: common between all tests. move to describe. Use `require.resolve('../../package.json')`, so it won't depend on the current cwd.

File: packages/cached/test/cached-fs-node.ix.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+        const fs = createCachedFs(nodeFs);
+
+        const stats = fs.statSync(filePath);
+        const stats2 = fs.statSync(filePath);
+
+        expect(stats).to.equal(stats2);
+    });
+
+    it('caches statsSync calls with invalidation', async () => {
```
Comment (AviVahl) [inline packages/cached/test/cached-fs-node.ix.ts]: it allows invalidating cache of a path

File: packages/cached/test/cached-fs-node.ix.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        const fs = createCachedFs(nodeFs);
+
+        const exists = fs.fileExistsSync(filePath);
+        const exists2 = fs.fileExistsSync(filePath);
+
+        expect(exists).to.equal(exists2);
+    });
+
+    it('caches stats (async) calls', async () => {
```
Comment (AviVahl) [inline packages/cached/test/cached-fs-node.ix.ts]: it caches stat (callback-style) calls

File: packages/cached/src/cached-fs.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            },
+            stat(path: string, callback: CallbackFn<IFileSystemStats>) {
+                path = createCacheKey(path);
+                const cachedStats = statsCache.get(path);
+                if (cachedStats) {
+                    callback(undefined, cachedStats);
+                    return;
+                }
+                fs.stat(path, (error: Error | null | undefined, stats: IFileSystemStats) => {
+                    statsCache.set(path, stats);
```
Comment (AviVahl) [inline packages/cached/src/cached-fs.ts]: avoid saving cache if there was an error

File: packages/cached/src/cached-fs.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+                return stats;
+            },
+            stat(path: string, callback: CallbackFn<IFileSystemStats>) {
+                path = createCacheKey(path);
+                const cachedStats = statsCache.get(path);
+                if (cachedStats) {
+                    callback(undefined, cachedStats);
+                    return;
+                }
+                fs.stat(path, (error: Error | null | undefined, stats: IFileSystemStats) => {
```
Comment (AviVahl) [inline packages/cached/src/cached-fs.ts]: are the explicit types really needed? inference doesn't work?

File: packages/cached/package.json (hunk 1 above)
Comment (tomermes) [reply packages/cached/package.json]: fixed

File: packages/cached/test/cached-fs-node.ix.ts (hunk 3 above)
Comment (tomermes) [reply packages/cached/test/cached-fs-node.ix.ts]: fixed

File: packages/cached/test/cached-fs-node.ix.ts (hunk 4 above)
Comment (tomermes) [reply packages/cached/test/cached-fs-node.ix.ts]: fixed

File: packages/cached/src/cached-fs.ts (hunk 6 above)
Comment (tomermes) [reply packages/cached/src/cached-fs.ts]: no, fixed

File: packages/cached/src/cached-fs.ts (hunk 5 above)
Comment (tomermes) [reply packages/cached/src/cached-fs.ts]: fixed

File: packages/cached/test/cached-fs-node.ix.ts (hunk 2 above)
Comment (tomermes) [reply packages/cached/test/cached-fs-node.ix.ts]: fixed

**Comment by tomermes** [comment]
> good work:
> 
> 1. wrap/test `promises.stat` as well.
> 2. add `invalidateAll()` method, for easier cleanup.

done

**Pull request merged**